tests/frontend-storage/cypress/integration/FrontendStorage.test.ts

```diff
@@ -2,34 +2,160 @@
+  });
+  
+  describe(`${frontendStorage.download.name}()`, () => {
+    for(const transferType of ["buffer", "stream"] as Array<FrontendTransferType>)
+    for(const inputMethod of ["url", "config"] as Array<InputMethod>) {
+      it(`should download to ${transferType} from ${inputMethod}`, async () => {
+        const testData = `test-download-to-${transferType}-using-${inputMethod}`;
+        const testDataBuffer = stringToArrayBuffer(testData);
+        const directory = await directoryManager.createNew();
+        const reference: ObjectReference = {
+          baseDirectory: directory.baseDirectory,
+          objectName: testData,
+        };
+        await serverStorage.upload({ data: testData, reference });
+
+        switch (inputMethod) {
+          case "url":
+            const url = await serverStorage.getDownloadUrl({ reference });
+            switch (transferType) {
+              case "buffer":
+                const responseBuffer = await frontendStorage.download({ url, transferType });
+                assertArrayBuffer(responseBuffer, testDataBuffer);
+                break;
+              case "stream":
+                const responseStream = await frontendStorage.download({ url, transferType });
+                await assertReadableStream(responseStream, testDataBuffer);
+                return;
+              default:
+                throw new Error(`Missing test case for transfer type ${transferType}`);
+            }
+            break;
+          case "config":
+            const transferConfig = await serverStorage.getDownloadConfig({ directory });
+            switch (transferType) {
+              case "buffer":
+                const responseBuffer = await frontendStorage.download({
+                  reference,
+                  transferConfig,
+                  transferType,
+                });
+                assertArrayBuffer(responseBuffer, testDataBuffer);
+                break;
+              case "stream":
+                const responseStream = await frontendStorage.download({
+                  reference,
+                  transferConfig,
+                  transferType,
+                });
+                await assertReadableStream(responseStream, testDataBuffer);
+                break;
+      
+              default:
+                throw new Error(`Missing test case for transfer type ${transferType}`);
+            }
+            break;
+            default:
+              throw new Error(`Missing test case for input method ${inputMethod}`);
+        }
+      });
+    }
   });
 
-  it("should download file from url", async () => {
-    const testFilePayload = "test payload";
-    const downloadUrl: string = await backendClient.getTestDownloadUrl(testFilePayload);
-    expect(downloadUrl).to.not.be.empty;
+  describe(`${frontendStorage.upload.name}()`, () => {
+    for(const transferType of ["buffer", "stream"] as  Array<FrontendTransferType>)
```

Comment:
I think the application of DRY here is a bit too extreme. The `for` loops here do a good job at covering the functionality but the readability is lost in the test itself. In this case the cyclomatic complexity of the test is very high, the loops not only determine which overload to call but also with what arguments. The switch statement in line 127 looks like at least two different tests to me as they test different workflows. 

The `ClientStorage.test.ts` has the following stucture which groups the tested methods by the workflow:
![image](https://user-images.githubusercontent.com/70565417/173332124-41aa0a17-1ac2-4c23-9675-54bae512dc47.png)

Maybe you can keep the variations of `reference` and `metadata` in test cases but I think the different workflows should be separated info different tests with common code moved into functions. Or you could maybe keep the data to upload/data type to download in test cases but at least separate the url transfer vs config transfer workflows.

Reply:
I refactored the switch statements into separate functions and moved them to test-templates, looks cleaner now thanks!